### ISSUE TITLE
RFC: stream category

### ISSUE THREAD

**Issue by kristianmandrup** [aws-amplify/amplify-cli#2614]
On a recent project we needed to trigger functions and notifications based on specific changes to the dynamoDB.

There was no such built-in support that we could find in the cli. We propose adding support for a new stream category:

- DynamoDB change streams
- Kinesis event streams

Currently we have a WIP [stream-category](https://github.com/kristianmandrup/amplify-cli/tree/stream-category) branch that could serve as an initial blueprint to be further improved.

So far we've had to add custom resources as per the docs and solutions to issues raised here

Cheers!

**Comment by kaustavghosh06** [comment]
@kristianmandrup Did you get a chance to take a look at https://github.com/aws-amplify/amplify-cli/pull/2463 ?

Currently we do allow trigger/streaming functionality for S3 and DDB tables added using the `amplify add storage` functionality. 
The above-mentioned PR enables you to add triggers/streaming to the DDB tables provisioned as a part of the GraphQL transformer as well.

**Comment by kristianmandrup** [comment]
Awesome!! However I find that the current CLI is hard to reason about as too much is wrapped into some of the categories.
Would be better with more fine grained categories IMO, and then with the option for some, such as storage to call into and extend using other categories, similar to how Yoman used to work with extensions/plugins.

Hope that PR you referenced gets merged real soon and the docs updated to reflect these additions, making it clear how to use it.

Our problem has been to understand how to extend existing resources and not just add them. Not very clear in the docs. Also had problems understanding how to work as a team on an amplify project.

Thanks.

**Comment by ambientlight** [comment]
categories might feel hard to reason about probably because the original intention was not have any tight-coupling to AWS services it feels, thus amplify provider can technically implement a non-AWS resource support, so categories are quite generic, personally for me, stream as a category feels intuitive in kinesis case, as part of #2463 kinesis stream will be available inside analytics category which might not be intuitive. 

Extending existing resources in ampllify usually  involves working with cloudformation templates directly, but it might get tricky when we depend on other categories (other cf stacks outputs), this is supported in amplify but currently not really explained in docs, yeah will be great to have docs explaining this.

**Comment by undefobj** [comment]
Thanks for the feedback @kristianmandrup we're waiting on the author of the PR to make some changes before merging. Please keep an eye on it. 

In general Categories are based on functional use cases and not implementations, and we try to not add them unless absolutely necessary. I think your requirements as outlined in the original post will be met by https://github.com/aws-amplify/amplify-cli/pull/2463.

Could you however expand with some concrete examples on a couple of your comments so that we could account for them in future designs specifically:
- storage to call into and extend using other categories, similar to how Yoman used to work with extensions/plugins
- how to extend existing resources and not just add them

**Comment by undefobj** [comment]
> the original intention was not have any tight-coupling to AWS services it feels, thus amplify provider can technically implement a non-AWS resource support

This is correct. Categories are functional use cases with declarative APIs that are non-AWS specific. It's also worth remembering that a Category is not just for the CLI but also for the entire Framework such as the JavaScript, iOS, Android libraries and UI components.

> Extending existing resources in ampllify usually involves working with cloudformation templates directly, but it might get tricky when we depend on other categories (other cf stacks outputs), this is supported in amplify but currently not really explained in docs, yeah will be great to have docs explaining this.

We do have some docs covering this but if there is more we could add here please let us know or add a PR:
- https://aws-amplify.github.io/docs/cli-toolchain/quickstart#custom-cloudformation-stacks
- https://aws-amplify.github.io/docs/cli-toolchain/graphql#custom-resolvers

That being said what @kaustavghosh06 and I have been recently discussing for the future is some easier ways to customize. This could potentially be automatically scaffolding out a custom stack or potentially alternatives. We're all ears and open to suggestions here.

**Comment by ambientlight** [comment]
> We do have some docs covering this but if there is more we could add here please let us know or add a PR

I was referring to the functionality where in `amplfy-meta.json` and in `backend-config.json`, dependsOn can be used to pass outputs of other category resources root stacks as parameters, or directly using this in parameters.json:

```
"Fn::GetAtt": ["stackName",  "Outputs.SomeOutput"]
```

**Comment by undefobj** [comment]
> > We do have some docs covering this but if there is more we could add here please let us know or add a PR
> 
> @undefobj: I was referring to the functionality where in `amplfy-meta.json` and in `backend-config.json`, `dependsOn` in category can be used to pass outputs of other category resources root stack outputs as parameters, or directly using this in parameters.json:
> 
> ```
> "Fn::GetAtt": ["stackName",  "Outputs.SomeOutput"]
> ```

@kaustavghosh06 @UnleashedMind thoughts on this?

**Comment by ambientlight** [comment]
> That being said what @kaustavghosh06 and I have been recently discussing for the future is some easier ways to customize. This could potentially be automatically scaffolding out a custom stack or potentially alternatives. 

@kaustavghosh06, @undefobj: 
can you guys expand on what `automatically scaffolding out of a custom stack` is?

> We're all ears and open to suggestions here.

As of now, `amplify-cli` has very little provider-independent logic, most of the categories logic will just relay CLI commands to the awscloudformation provider that would perform CLI flow, emitting the resulting CF templates. Rolling out another provider basically would be sort of reimplementing the functionality of each category inside amplify-cli scope, curious if you guys thought of moving some of that logic provider-independent: i.e very roughly: cli flows can emit intermediate representations that are provider-agnostic, those can bundle cross-category dependencies, permission classes, etc and then passed to provider to resolve. 

On the client side we would then be able to create instances implementing `Provider` interface that would convert intermediate representation to any `template` representation like `CF` or use imperative approaches akin to [aws-cdk](https://github.com/aws/aws-cdk) or [serverless/components](https://github.com/serverless/components). 

That can be also tricky in terms of those magical graphQL `@transformers` where `graphql-transformer-core` is tighly-coupled to CF. (But IMHO, this appsync abstraction is a single-for-now-unique feature of `amplify-cli` that is not available in other venture-backed infra-as-code solution like `serverless-framework / terraform / pulumi`)

Curious to hear your thoughts on this.

**Comment by kaustavghosh06** [comment]
@kristianmandrup We recently merged @ambientlight 's PR for kinesis and DDB streams (DDB generated as a part of the GraphQL transformer)
 You can checkout the documentation out here - https://aws-amplify.github.io/docs/cli-toolchain/quickstart#as-a-part-of-the-graphql-api-types-with-model-annotation and https://aws-amplify.github.io/docs/cli-toolchain/quickstart#kinesis-stream-trigger

@ambientlight For the provider plugin hooks, could you maybe open a separate issue or an RFC since it seems to be divulging from the original issue.

**Comment by kristianmandrup** [comment]
@kaustavghosh06 Great to see!

"You can also associated a Lambda trigger" should be "You can also associate a Lambda trigger".

"After the above question, you can select one of the types annotated by @model for which you want to add a trigger for."

Please show an example of this in the docs, ie. selecting `eventID` and `eventName` I would think?

```js
    console.log(record.eventID);
    console.log(record.eventName);
```

`record.dynamodb` will contain a DynamoDB change json describing the item changed in DynamoDB table.

Please show an example of such a JSON change document.

You can add a Kinesis stream to your Amplify project by going through the amplify add analytics flow

I don't understand why Kinesis is under analytics. I would think it is core infrastructure for decoupled micro services in general? Why not an `eventing` category or sth?

Otherwise everything looks great to me :) Good job!
